curp/src/server.rs

```diff
@@ -621,93 +655,79 @@ impl<C: 'static + Command, CE: 'static + CommandExecutor<C>> Protocol<C, CE> {
-                                            &ProposeError::ExecutionError(err.to_string()),
-                                        )
-                                    },
-                                    |rv| {
-                                        ProposeResponse::new_result::<C>(is_leader, self_term, &rv)
-                                    },
-                                )
-                            })
-                            .right_future()
-                            .left_future()
-                    }
-                    // not conflict and is not leader
-                    (false, false) => {
-                        async { ProposeResponse::new_empty(false, self.state.read().term) }
-                            .right_future()
-                            .right_future()
+
+        (|| async {
+            let (is_leader, term) = self.state.map_read(|state| (state.is_leader(), state.term));
+            let sync_notify = {
+                let mut spec = self.spec.lock();
+                let has_conflict = spec.has_conflict_with(&cmd);
+
+                if !has_conflict {
+                    spec.push(cmd.clone());
+                }
+
+                // non-leader should return immediately
+                if !is_leader {
+                    return if has_conflict {
+                        ProposeResponse::new_error(is_leader, term, &ProposeError::KeyConflict)
+                    } else {
+                        ProposeResponse::new_empty(false, term)
+                    };
+                }
+
+                // the leader will sync the command to others while grabbing the lock so that the order of the command can be preserved
+                let sync_notify = match self.sync_to_others(term, &cmd, true) {
+                    Ok(notify) => notify,
+                    Err(err) => {
+                        return ProposeResponse::new_error(
+                            is_leader,
+                            self.state.read().term,
+                            &ProposeError::SyncedError(format!(
+                                "Sync cmd channel closed, {:?}",
+                                err
+                            )),
+                        );
                     }
+                };
+
+                // if the command has conflict, sync immediately and return conflict error
+                if has_conflict {
+                    sync_notify.notify(1);
+                    return ProposeResponse::new_error(is_leader, term, &ProposeError::KeyConflict);
                 }
-            })
-            .await
-            .map(tonic::Response::new)
-            .map_err(|err| tonic::Status::internal(format!("encode or decode error, {}", err)))
+
+                // finally, the command meets all the conditions for speculative execution, we will release the lock here
+                sync_notify
+            };
+
+            // speculative execution
```

Comment:
Add an issue to optimize the sync channel blocking mechanism. For instance, we can provide an execution task to serializing the execution orders.